### ISSUE TITLE
[Codegen] Attach `!range` to slice/array length loads so a length is known non-negative

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1080]
Baselines: Align `8c8bfbc7a3169e84ecc8415f5149ab8c61afe863` (alignc 0.7.5, LLVM 22.1.8), align-llm
`b72b0b48`, Apple M1, `--profile release`, default `--target-cpu baseline`.

## Problem

A slice/array `len` field is non-negative by construction, but the load that materializes it carries no
metadata, so LLVM has to assume it can be negative. Every loop bounded by `len` then pays a
`llvm.smax(len, 0)` clamp, every two-length loop additionally pays `llvm.umin`, and the vectorizer picks a
narrower width because the trip count is not provably positive.

This is small, self-contained, and one of the three parts of a single **loop-facts contract** (see
*Relation* below).

## Evidence

Instruction metadata the compiler emits today, whole program:

```
$ git grep -n -E '!range|llvm\.assume' -- 'crates/**/*.rs'     # 0 hits
```

| fact | count in the 746,869-line whole-program optimized IR of align-llm |
|---|---|
| `!range` on any load | **0** |
| `llvm.assume` | **0** |
| `call i64 @llvm.smax.i64` | **536** |
| `call i64 @llvm.umin.i64` | **33** |

The clamp is visible in the shipped machine code. `kv_plane$all_zero`, verbatim from
`otool`/`objdump` of the release image:

```
000000010004bf84 <_align_fn$17$6b765f706c616e6524616c6c5f7a65726f>:
10004bf84:     	ldp	x9, x8, [x0]
10004bf88:     	mov	x12, xzr
10004bf8c:     	bic	x10, x8, x8, asr #63    ; = smax(len, 0)
<L0>:
10004bf90:     	mov	x11, x12
10004bf94:     	cmp	x10, x12
10004bf98:     	b.eq	 <L1>
10004bf9c:     	ldrb	w13, [x9, x11]
10004bfa0:     	add	x12, x11, #0x1
10004bfa4:     	cbz	w13,  <L0>
```

`bic xD, xN, xN, asr #63` is the `smax(len, 0)` idiom; it occurs **351** times in the release image —
every one with identical source registers, i.e. every one of them this clamp.

The same emitted `{ptr,len}` header load shows up in the raw IR as:

```llvm
  %load1.elt10 = getelementptr inbounds nuw i8, ptr %0, i64 8
  %load1.unpack11 = load i64, ptr %load1.elt10, align 8
  %smax = tail call i64 @llvm.smax.i64(i64 %load1.unpack11, i64 0)
```

**Measured effect of supplying the fact.** Asserting `len >= 0` once per header materialization (5 sites)
in the *emitted raw IR only* and re-running the identical `default<O2>` pipeline with `llvm@22/bin/opt`
on an `f32` `dot(a, b)` kernel:

| | `smax`/`umin` guards | vector width |
|---|---|---|
| as emitted | 8 | `<2 x float>` |
| `len >= 0` known | **0** | **`<4 x float>`** |

Independently confirmed on real client code: `llvm.smax` appears 5× in
`moe_decode_step$run_moe_layer`, 8× in `tokenizer_qwen2$encode_raw_append`, 17× in
`decode_step$execute_mode_traced` — all per-token or per-token-per-layer paths.

## Root cause

Codegen emits the `len` field load with no metadata at all
(`crates/align_codegen_llvm/src/lib.rs:22357-22376`, `borrowed_view_part`; the `extractvalue` path for
by-value headers in `Rvalue::SliceLen`). LLVM has no other source for the invariant: the header is
ordinary memory to it.

## Proposed solution

Attach `!range !{i64 0, i64 <max>}` to every load that materializes a slice/array length — one metadata
attachment at the single place where the `len` field is read, keyed off the field index already known to
`borrowed_view_part`. Nothing else changes; the bound is the language's own capacity limit.

This is deliberately **not** the rejected `llvm.assume` policy. `docs/open-questions.md:4187-4189` lists
*"`llvm.assume` / early intrinsic emission / loop-metadata overrides as a general policy"* among the
rejected directions, with the reason *"attributes and flags first"*. `!range` on a load is exactly an
"attributes and flags first" mechanism: it is per-load, carries no control dependence, costs no
instruction, and cannot be hoisted into a place where it asserts something false. No document in `docs/`
mentions `!range` at all, so nothing owns or forbids it today.

For consistency the same treatment applies to every length source, so there is one rule rather than a
special case: borrowed header loads, by-value header `extractvalue`, and fixed-array constant lengths
(where the value is already a constant and the metadata is a no-op).

## Acceptance criteria

- `!range !{i64 0, ...}` appears on the length load in emitted IR for a borrowed `slice<T>`, a borrowed
  `array<T>`, and a by-value slice parameter.
- An `f32` `dot(a, b)` kernel emits **0** `llvm.smax`/`llvm.umin` guards and a `<4 x float>` vector body
  at `--profile release --target-cpu baseline` (today: 8 guards, `<2 x float>`).
- `bic xD, xN, xN, asr #63` count in the align-llm release image drops from 351 toward 0; whole-program
  `llvm.smax.i64` call sites drop from 536.
- `kv_plane$all_zero` loses its `bic` clamp.
- A codegen owner test pins the metadata on the length load; existing
  `crates/align_driver/tests/vectorize_shapes.rs` passes unchanged.
- No change to any trap/diagnostic: the `(index, len)` text reported by `align_rt_bounds_fail` /
  `align_rt_range_fail` stays byte-identical.

## Relation to existing issues/plans

Part 2 of the **loop-facts contract** — three changes that together determine what an Align loop tells
LLVM about its own data, and that should be designed as one contract:

1. header materialization and alias facts — *"[Codegen] Materialize a borrowed slice/array view
   header once and state its alias facts"* (#1079);
2. **this issue** — length facts;
3. check shape — *"[MIR] Fuse the bounds check into one unsigned compare and eliminate
   monotone-induction in-loop checks"* (#1081).

`!range` is the cheapest of the three and the only one that is purely additive.
`docs/impl/63-codegen-performance-audit.md:229-231` already records that *"checking
`src.len() >= 0`, or expressing the loop limit with unsigned casts, enables a vector main loop for
the integer byte-sum witness"* — i.e. the payoff was observed from the caller side and left to the
caller; this issue moves it into the compiler so callers do not have to write the condition. That
plan's own guidance at `:233-234` is to *"use the valid-view length invariant in the proof owner
rather than requiring callers to write these extra conditions"*.

Not a duplicate of #1064 (typed byte views), #1066 (`--rt-lto`/ThinLTO inlining) or #1063 (`exp`).

## Impact framing

Two instructions and a narrowed vector width per loop; on its own a few per cent, not a headline number
(in a real Metal generation profile Align-compiled code is ~0.1 % of CPU during GPU decode). Its value is
(a) it doubles the achievable vector width on the two-length kernels that the rest of the loop-facts
contract unlocks, (b) it is measurable on the Align-dominated tokenizer path (74.5 % Align code) and the
per-token CPU kernels (greedy logits selection 242 µs/token over 151,936 logits, sampler 267 µs/token
over 152,064, both on this host), and (c) 351
redundant clamp instructions in a 5.3 MB binary is a code-quality statement for the general "Align must
emit fast binaries" goal.

**Comment by sanohiro** [comment]
align-llm consumer residual at managed Align d9b0df32 (macOS arm64, release, no rt-LTO): the named kv_plane.all_zero witness still materializes its borrowed slice length without !range. The exact optimized core begins with load ptr from the header, then load i64 from header+8, then llvm.smax.i64(len, 0). The shipped product image therefore retains `bic x10, x8, x8, asr #63` in kv_plane$all_zero; the whole-image exact bic clamp count is 13 versus the original 351 baseline. Replacing only view.u8(at) with direct view[at] removes no residual: the exported unit core still has the unannotated length load and smax. This is a contract-admitted borrowed slice parameter, so it appears to miss Plan 69 PR1 rather than being an application fallback. tokenizer and alignpack smokes pass; no Linux measurement was run.

**Comment by sanohiro** [comment]
The named `kv_plane.all_zero` residual is fixed at Align
`5c7af9e54108fbe3a7b3d698c96a6dbc2da3e9c3` (includes #1166/#1168):
the actual client unit now has a nonnegative borrowed length, no `llvm.smax`,
and a `<16 x i8>` vector loop.

A second borrowed-header path still misses Plan 69 I7. Adding an independent
one-element local allocation to the same scan disables cached view headers;
the fallback aggregate load then loses the length fact as well:

```align
module range_residual

pub fn plain(borrow view: slice<u8>) -> bool {
  mut at := 0
  loop {
    if at >= view.len() { break }
    if view.u8(at) as i64 != 0 { return false }
    at = at + 1
  }
  return true
}

pub fn allocated(borrow view: slice<u8>) -> bool {
  scratch := [0].to_array()
  mut at := 0
  loop {
    if at >= view.len() { break }
    if view.u8(at) as i64 != 0 { return false }
    at = at + 1
  }
  return scratch.len() == 1
}
```

Reproduce using the release compiler at the exact revision above:

```sh
alignc emit-llvm range-residual.align --stage raw --profile release --no-rt-lto
alignc emit-llvm range-residual.align --stage optimized --profile release --no-rt-lto
```

Observed on Apple M1, macOS 27.0, LLVM 22.1.8, baseline target:

| Function | Raw borrowed header | Optimized result |
| --- | --- | --- |
| `plain` | scalar length load with `!range !{i64 0, i64 -9223372036854775808}` | vector body; no `llvm.smax` |
| `allocated` | `load { ptr, i64 }, ptr %0` followed by `extractvalue`; no range fact | unannotated scalar length load; `llvm.smax.i64(len, 0)`; scalar scan |

The real `runtime_sampler.select_values` consumer also reloads its borrowed
`slice<f32>` length inside the vocabulary loop without `!range`.

At this revision, `ViewFactsPlan` explicitly says `!range` is independent of
its whole-body alias/caching gate. `load_view_part` implements that rule, but
`Rvalue::Load` at `crates/align_codegen_llvm/src/lib.rs:13885` falls back to an
aggregate load when `cached_view_header` is unavailable. Its later
`SliceLen(Value)` uses `extractvalue`, bypassing `load_view_part`. The
allocation fails the whole-body whitelist, exposing this path.

The demonstrated defect is the missing unconditional length invariant.
The absence of `noalias`/header caching is separately permitted by the
current conservative whole-body policy. Restoring the range fact alone is
not claimed to guarantee SIMD or to remove all sampler reloads.

Suggested acceptance: preserve the independently justified nonnegative length
fact on typed borrowed-header materialization even when caching is disabled,
cover the example above and the real sampler, and retain the existing
non-length-layout and foreign-provenance negative controls. No syntax, unsafe
assumption, or application spelling workaround is needed.

The exact fixture, measurement record and reproduction notes are merged in
[align-llm PR #293](https://github.com/sanohiro/align-llm/pull/293), specifically
[`range-residual.align`](https://github.com/sanohiro/align-llm/blob/0d9eb8736c7805968046d638b25185745684696f/eval/benchmarks/codex-binary-audit-2026-09-23/range-residual.align)
and the [audit report](https://github.com/sanohiro/align-llm/blob/0d9eb8736c7805968046d638b25185745684696f/eval/benchmarks/codex-binary-audit-2026-09-23/README.md).

Correction to my earlier comment on this issue: the old image's previously
reported whole-image syntactic clamp count of 13 does not reproduce under the
now checked-in explicit predicate. The identified old and new images count
50 and 45 respectively. These are syntactic counts, not an attribution of
every clamp to a borrowed view. The `kv_plane.all_zero` repair and the
uncached-header residual above have direct IR evidence independent of that
aggregate count.

**Comment by sanohiro** [comment]
Align correction merged in PR #1170 (`829e1da828a6295c31bcabd4deaebcf93e3036cf`). The uncached `Rvalue::Load` path now reads typed borrowed `{ptr, len}` header fields separately at each use, so the scalar length load receives `!range` even when the independent whole-body cache/alias proof fails. The change does not enable caching, TBAA, or `noalias` for the allocation case; `json.doc` and other non-length `{ptr, i64}` layouts remain excluded.

I reproduced the reported `plain`/`allocated` contrast locally before the fix. With the merged compiler, the exact allocation-plus-scan fixture has `!range` on its raw and optimized borrowed length loads and no `llvm.smax.i64` in the `allocated` function. The real `runtime_sampler.select_values` source also compiles locally with `!range` on its borrowed `slice<f32>` length loads in raw and optimized IR. Its optimized IR still has an unrelated `smax(x, 1)`; this fix makes no whole-image clamp-count or throughput claim. The `loop_facts` owner (45 tests), bounded gate, Clippy, and Linux x86_64/ARM64/macOS CI pass.

The Align provider defect described by this reopening is fixed, so I am closing this issue. Request 108 remains `ALIGN_MERGED` until consumer-owned final adoption/measurement; no additional align-llm task is requested here.
